### ISSUE TITLE
Exception: java.lang.OutOfMemoryError due to a huge text file 

### ISSUE THREAD

**Issue by HanzhiWang1105** [mmolimar/kafka-connect-fs#32]
Hi, I get the same error "Exception: java.lang.OutOfMemoryError" when to send a 28Mb text file. 
Please find all details, 
```
[2018-05-17 15:07:44,292] INFO Processing records for file [path = file:/tmp/testFile/t.txt, length = 29288895, blocks = [[offset = 0, length = 29288895, corrupt = false]]] (com.github.mmolimar.kafka.connect.fs.FsSourceTask:73)

Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "kafka-producer-network-thread | producer-1"

Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "pool-1-thread-1"

Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "org.eclipse.jetty.server.session.HashSessionManager@48ef6327Timer"

Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "pool-3-thread-1"
```
I would like to know the limit of size of file. Thank you a lot.

[](https://github.com/mmolimar/kafka-connect-fs/blob/master/src/main/java/com/github/mmolimar/kafka/connect/fs/FsSourceTask.java)

**Comment by mmolimar** [comment]
I've processed bigger files than 28 Mb.
Try increasing the heap memory in the JVM.

**Comment by HanzhiWang1105** [comment]
@mmolimar  Thanks a lot for your reply.
It works if I increase the heap memory. Does it means the size of file is only up to heap memory? 
If I want to send a GB file or larger file, is increasing heap memory enough to do it?
Thanks again for your reply.

**Comment by mmolimar** [comment]
No, it doesn't. It's all about how the JVM handles the memory and manages the GC.

**Comment by HanzhiWang1105** [comment]
@mmolimar When you said "How the JVM handles the memory and manages the GC", it means in your code you have already considered it or should I add this part to realize the goal?
I have little experience and sorry to ask question like this.
Thanks.

**Comment by HanzhiWang1105** [comment]
@mmolimar Now I totally find the issue with the tool Memory Analyzer Tool .
In the class "FsSourceTask.java" method poll() line74-76, the arraylist results will add all records of the whole file and return. If I have a large file, it will not finish but read all lines. Therefore it causes the heap memory out.

**Comment by hasnat** [comment]
@mmolimar should we be batching this with some config e.g.
https://github.com/mmolimar/kafka-connect-fs/blob/master/src/main/java/com/github/mmolimar/kafka/connect/fs/FsSourceTask.java#L75
```diff
                   while (reader.hasNext()) {
                        results.add(convert(metadata, reader.currentOffset(), reader.next()));
+                        if (results.size() < config.get(RECORDS_BATCH_LIMIT)) {
+                            log.info("Reached batch limit for records {}", metadata);
+                            break;
                        }
                    }
```
not sure if we should limit total records (regardless of file) or records by file

**Comment by shengwu1026** [comment]
@mmolimar @HanzhiWang1105 Could you please tell me how to change the heap memory? I tried java -Xmx, but I was not sure which jar file or class I should put in the command.

**Comment by kylecbrodie** [comment]
@HanzhiWang1105 @mmolimar I believe this issue is resolved in version `1.1.0` with the config param `file_reader.batch_size`
